utils/emmylua/wesnoth/units.lua

```diff
@@ -0,0 +1,265 @@
+---@field upkeep integer|"'full'"|"'free'"|"'loyal'"
+---@field usage string
+---@field renamable boolean
+---@field undead_variation string
+---@field role string
+---@field resting boolean
+---@field recall_filter WML
+---@field extra_recruit string[]
+---@field advances_to string[]
+---@field advancements WMLTable[]
+---@field status table<string, boolean>
+---@field variables table<string, WML>
+---@field attacks unit_weapon[]
+---@field traits string[]
+---@field abilities string[]
+---@field animations string[]
+---@field __cfg WMLTable
+
+---@class wesnoth.units
+wesnoth.units = {}
+
+---Advances the unit if has enough XP_attack
+---@param unit unit
+---@param animate? boolean
+---@param fire_events? boolean
+function wesnoth.units.advance(unit, animate, fire_events) end
+
+---Creates a copy of the unit
+---@param unit unit
+function wesnoth.units.clone(unit) end
+
+---Erases the unit from the map
+---@param unit unit
+---@overload fun(loc:location)
+---@overload fun(x:integer, y:integer)
+function wesnoth.units.erase(unit) end
+
+---Extracts the unit from the map, making it a private unit
+---@param unit unit
+function wesnoth.units.extract(unit) end
+
+---Tests if the unit matches a filter
+---@param unit unit The unit to test
+---@param filter WML The filter to match
+---@param context? location|unit A second unit for the filter, or a reference location
+function wesnoth.units.matches(unit, filter, context) end
+
+---Place or move a unit on the map
+---@param unit unit
+---@param loc? location
+---@overload fun(unit:unit, x:integer, y:integer)
+function wesnoth.units.to_map(unit, loc) end
+
+---Place a unit on a recall lists
+---@param unit unit
+---@param side? integer
+function wesnoth.units.to_recall(unit, side) end
+
+---Transform a unit into a different type
+---@param unit unit
+---@param to_type string
+---@param to_variation? string
+function wesnoth.units.transform(unit, to_type, to_variation) end
+
+---Select the unit, as if it had been clicked with the mouse
+---@param unit unit
+function wesnoth.units.select(unit) end
+
+---Test if the unit is affected by an ability
+---@param unit unit
+---@param tag_name string
+---@return boolean
+function wesnoth.units.ability(unit, tag_name) end
+
+---Get the defense of a unit on a particular terrain
+---@param unit unit
+---@param terrain string
+---@return number
+function wesnoth.units.defense_on(unit, terrain) end
+
+---Get the resistance of a unit againt an attack type
+---@param unit unit The unit to check
+---@param damage_type string The damage type
+---@param as_attacker? boolean Whether to consider the unit as the attacker or the defender
+---@param location? location Calculate as if the unit is on this location
+---@return number
+---@overload fun(unit:unit, damage_type:string, as_attacker?:boolean, x:integer, y:integer)
+function wesnoth.units.resistance_against(unit, damage_type, as_attacker, location) end
+
+---Get the movement of a unit on a particular terrain
```

Comment:
Might be clearer as "get the movement cost for a unit on a particular terrain".

Similarly for vision and jamming.